examples/cvode/ginkgo/cv_heat2D_ginkgo.cpp

```diff
@@ -0,0 +1,587 @@
+// excluding the corners (set by J_sn_kernel).
+__global__ void J_we_kernel(const sunindextype nx, const sunrealtype ny, sunindextype* row_ptrs, sunindextype* col_idxs,
+                            sunrealtype* mat_data)
+{
+  const sunindextype j = blockIdx.x * blockDim.x + threadIdx.x;
+
+  if (j > 0 && j < ny - 1) {
+    // Western face
+    auto col      = j * nx;
+    auto idx      = (5 * (nx - 2) + 2) * (j - 1) + nx;
+    mat_data[idx] = ZERO;
+    col_idxs[idx] = col;
+    row_ptrs[col] = idx;
+
+    // Eastern face
+    col           = (nx - 1) + j * nx;
+    idx           = (5 * (nx - 2) + 2) * (j - 1) + nx + 1 + 5 * (nx - 2);
+    mat_data[idx] = ZERO;
+    col_idxs[idx] = col;
+    row_ptrs[col] = idx;
+  }
+}
+
+// GPU kernel to compute the ODE RHS Jacobian function df/dy(t,y).
+__global__ void J_kernel(const sunindextype nx, const sunindextype ny, const sunrealtype cx, const sunrealtype cy,
+                         const sunrealtype cc, sunindextype* row_ptrs, sunindextype* col_idxs, sunrealtype* mat_data)
+{
+  const sunindextype i = blockIdx.x * blockDim.x + threadIdx.x;
+  const sunindextype j = blockIdx.y * blockDim.y + threadIdx.y;
+
+  if (i > 0 && i < nx - 1 && j > 0 && j < ny - 1) {
+    auto row   = i + j * nx;
+    auto col_s = row - nx;
+    auto col_w = row - 1;
+    auto col_c = row;
+    auto col_e = row + 1;
+    auto col_n = row + nx;
+
+    // Number of non-zero entries from preceding rows
+    auto prior_nnz = (5 * (nx - 2) + 2) * (j - 1) + nx;
+
+    // Starting index for this row
+    auto idx = prior_nnz + 1 + 5 * (i - 1);
+
+    mat_data[idx]     = cy;
+    mat_data[idx + 1] = cx;
+    mat_data[idx + 2] = cc;
+    mat_data[idx + 3] = cx;
+    mat_data[idx + 4] = cy;
+
+    col_idxs[idx]     = col_s;
+    col_idxs[idx + 1] = col_w;
+    col_idxs[idx + 2] = col_c;
+    col_idxs[idx + 3] = col_e;
+    col_idxs[idx + 4] = col_n;
+
+    row_ptrs[row] = idx;
+  }
+}
+#endif
+
+// J routine to compute the ODE RHS Jacobian function df/dy(t,y). This
+// explicitly set boundary entries to zero so J(t,y) has the same sparsity
+// pattern as A = I - gamma * J(t,y).
+int J(sunrealtype t, N_Vector y, N_Vector fy, SUNMatrix J, void* user_data, N_Vector tmp1, N_Vector tmp2, N_Vector tmp3)
+{
+  // Access problem data
+  auto udata    = static_cast<UserData*>(user_data);
+  const auto nx = udata->nx;
+  const auto ny = udata->ny;
+  const auto dx = udata->dx;
+  const auto dy = udata->dy;
+  const auto kx = udata->kx;
+  const auto ky = udata->ky;
+
+  // Constants for computing J(t,y)
+  const sunrealtype cx = kx / (dx * dx);
+  const sunrealtype cy = ky / (dy * dy);
+  const sunrealtype cc = -TWO * (cx + cy);
+
+  auto J_gko = static_cast<SUNGkoMatrixType*>(J->content)->GkoMtx();
+
+  sunindextype* row_ptrs = J_gko->get_row_ptrs();
+  sunindextype* col_idxs = J_gko->get_col_idxs();
+  sunrealtype* mat_data  = J_gko->get_values();
+
+#if defined(USE_CUDA) || defined(USE_HIP)
+
+  unsigned threads_per_block_bx = 16;
```

Comment:
Why 16 @gardner48 ?

Reply:
Copy-paste error from the `{16,16}` blocks used below